src/pinata/mod.rs

```diff
@@ -0,0 +1,59 @@
+use std::{env, path::Path};
+
+use dotenvy::dotenv;
+use pinata_sdk::{PinByFile, PinataApi};
+
+use crate::upload::UploadError;
+
+pub trait PinataClient: Sized {
+    fn new() -> impl std::future::Future<Output = Result<Self, UploadError>> + Send;
+    fn upload_file_to_ipfs(
+        &self,
+        path: &Path,
+    ) -> impl std::future::Future<Output = Result<String, UploadError>> + Send;
+}
+
+pub struct PinataClientImpl {
+    pinata_api: PinataApi,
+}
+
+impl PinataClient for PinataClientImpl {
+    async fn new() -> Result<Self, UploadError> {
+        dotenv().ok();
+        match (env::var("PINATA_API_KEY"), env::var("PINATA_API_SECRET")) {
+            (Ok(api_key), Ok(secret_api_key)) => {
+                let api = PinataApi::new(api_key, secret_api_key)
+                    .map_err(|_| UploadError::Authentication)?;
+                api.test_authentication()
+                    .await
+                    .map_err(|_| UploadError::Authentication)?;
+                Ok(PinataClientImpl { pinata_api: api })
+            }
+            _ => Err(UploadError::Ipfs),
+        }
```

Comment:
nit: reduced nesting

```suggestion
    dotenv().ok();
    
    let (api_key, secret_api_key) = match (env::var("PINATA_API_KEY"), env::var("PINATA_API_SECRET")) {
        (Ok(key), Ok(secret)) => (key, secret),
        _ => return Err(UploadError::Ipfs),
    };

    let api = PinataApi::new(api_key, secret_api_key)
        .map_err(|_| UploadError::Authentication)?;
        
    api.test_authentication()
        .await
        .map_err(|_| UploadError::Authentication)?;
        
    Ok(PinataClientImpl { pinata_api: api })
```

Reply:
I like both approaches 😅

With Sophie's it's compact which I'm a big fan of.

With Zeeshan's it's return early which I also a big fan of, but I've recently starting trying to not leave variables dangling for the entire scope that aren't used (most of the time it's hard to do, but it's just a style that I've picked up lately).

Combining both approaches:

```
        let api = match (env::var("PINATA_API_KEY"), env::var("PINATA_API_SECRET")) {
            (Ok(api_key), Ok(secret_api_key)) => {
                PinataApi::new(api_key, secret_api_key)
                    .map_err(|_| UploadError::Authentication)?
            }
            _ => return Err(UploadError::Ipfs),
        };

        api.test_authentication()
            .await
            .map_err(|_| UploadError::Authentication)?;

        Ok(PinataClientImpl { pinata_api: api })
```

Reply:
done in f6f0382a6b254a82d9937247e06cf822c4245520